ext/random/randomizer.c

```diff
@@ -243,6 +243,90 @@ PHP_METHOD(Random_Randomizer, pickArrayKeys)
 }
 /* }}} */
 
+/* {{{ Get Random Bytes for String */
+PHP_METHOD(Random_Randomizer, getBytesFromString)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_long length;
+	zend_string *source, *retval;
+	size_t total_size = 0;
+
+	ZEND_PARSE_PARAMETERS_START(2, 2);
+		Z_PARAM_STR(source)
+		Z_PARAM_LONG(length)
+	ZEND_PARSE_PARAMETERS_END();
+
+	const size_t source_length = ZSTR_LEN(source);
+
+	if (source_length < 1) {
+		zend_argument_value_error(1, "cannot be empty");
+		RETURN_THROWS();
+	}
+
+	if (length < 1) {
+		zend_argument_value_error(2, "must be greater than 0");
+		RETURN_THROWS();
+	}
+
+	retval = zend_string_alloc(length, 0);
+
+	if (source_length > 0xFF) {
+		while (total_size < length) {
+			uint64_t offset = randomizer->algo->range(randomizer->status, 0, source_length - 1);
+
+			if (EG(exception)) {
+				zend_string_free(retval);
+				RETURN_THROWS();
+			}
+
+			ZSTR_VAL(retval)[total_size++] = ZSTR_VAL(source)[offset];
+		}
+	} else {
+		uint64_t mask = 0xFF;
+		if (source_length <= 0x7F) mask = 0x7F;
+		if (source_length <= 0x3F) mask = 0x3F;
+		if (source_length <= 0x1F) mask = 0x1F;
+		if (source_length <= 0xF) mask = 0xF;
+		if (source_length <= 0x7) mask = 0x7;
+		if (source_length <= 0x3) mask = 0x3;
+		if (source_length <= 0x1) mask = 0x1;
```

Comment:
There's an off-by-one in the masking logic: The result is not incorrect, it's just less efficient.

If the input length is of length `4`, then we may index the indices 0 to 3, i.e. `0b00`, `0b01`, `0b10`, `0b11`. However the current implementation selects the `0x7` mask which is `0b111` and thus contains an additional useless bit.

Also I think this would be more readable with `else if`. I suggest:

```
  uint64_t mask;

  if (source_length <= 0x1)
    mask = 0x0;
  else if (source_length <= 0x2)
    mask = 0x1;
  else if (source_length <= 0x4)
    mask = 0x3;
  else if (source_length <= 0x8)
    mask = 0x7;
  else if (source_length <= 0x10)
    mask = 0xF;
  else if (source_length <= 0x20)
    mask = 0x1F;
  else if (source_length <= 0x40)
    mask = 0x3F;
  else if (source_length <= 0x80)
    mask = 0x7F;
  else
    mask = 0xFF;
```

There's likely also some magic possible with `__builtin_clz`, but that's likely not worth it here and optimizing this later is always possible.